src/Coordination/KeeperStorage.cpp

```diff
@@ -1344,6 +1362,212 @@ struct KeeperStorageSetRequestProcessor final : public KeeperStorageRequestProce
+    for (const RaftServerConfig & update : parseRaftServers(joining))
+        if (auto server_ptr = config->get_server(update.id))
+        {
+            if (update.endpoint != server_ptr->get_endpoint()
+                || update.learner != server_ptr->is_learner()
+                || update.priority == server_ptr->get_priority())
+                return {}; // can't change server endpoint/type due to NuRaft API limitations
+            out.emplace_back(UpdateRaftServerPriority{.id = update.id, .priority = update.priority});
+        }
+        else if (endpoints.contains(update.endpoint))
+            return {};
+        else
+            out.emplace_back(AddRaftServer{update});
+
+    return out;
+}
+
+ClusterUpdateActions processLeaving(const ClusterConfigPtr & cfg, std::string_view leaving)
+{
+    std::vector<std::string_view> leaving_arr;
+    splitInto<','>(leaving_arr, leaving);
+    if (leaving_arr.size() >= cfg->get_servers().size())
+        return {};
+
+    std::unordered_set<int> remove_ids;
+    ClusterUpdateActions out;
+
+    for (std::string_view leaving_server : leaving_arr)
+    {
+        int id;
+        if (std::from_chars(leaving_server.begin(), leaving_server.end(), id).ec != std::error_code{})
+            return {};
+
+        if (remove_ids.contains(id))
+            continue;
+
+        if (auto ptr = cfg->get_server(id))
+            out.emplace_back(RemoveRaftServer{.id = id});
+        else
+            return {};
+
+        remove_ids.emplace(id);
+    }
+
+    return out;
+}
+
+String serializeClusterConfig(const ClusterConfigPtr& config, const ClusterUpdateActions& updates)
+{
+    RaftServers new_config;
+    std::unordered_set<int> remove_update_ids;
+
+    for (const auto& update : updates)
+    {
+        if (const auto* add = std::get_if<AddRaftServer>(&update))
+            new_config.emplace_back(*add);
+        else if (const auto* remove = std::get_if<RemoveRaftServer>(&update))
+            remove_update_ids.insert(remove->id);
+        else if (const auto* priority = std::get_if<UpdateRaftServerPriority>(&update))
+        {
+            remove_update_ids.insert(priority->id);
+            new_config.emplace_back(RaftServerConfig{*config->get_server(priority->id)});
+        }
+        else
+            UNREACHABLE();
+    }
+
+    for (const auto& item : config->get_servers())
+        if (!remove_update_ids.contains(item->get_id()))
+            new_config.emplace_back(RaftServerConfig{*item});
+
+    return fmt::format("{}", fmt::join(new_config.begin(), new_config.end(), "\n"));
+}
+
+std::vector<KeeperStorage::Delta> KeeperStorageReconfigRequestProcessor::preprocess(
+    KeeperStorage & storage, int64_t zxid, int64_t, int64_t time, uint64_t & digest, const KeeperContext & ctx) const
+{
+    using enum Coordination::Error;
+    ProfileEvents::increment(ProfileEvents::KeeperReconfigRequest);
+
+    if (!ctx.dispatcher->reconfigEnabled())
+        return {{zxid, ZUNIMPLEMENTED}};
```

Comment:
ZK C client has a special return code for that https://github.com/search?q=repo%3Aapache%2Fzookeeper%20zreconfigdisabled&type=code, although server just throws a NonImplemented error